test/vinyl-luatest/update_optimize_test.lua

```diff
@@ -0,0 +1,407 @@
+    t.assert_equals(box.snapshot(), 'ok')
+    dump_count = wait_for_dump(index, dump_count)
+    new_stmt_count = dumped_stmt_count({index, index2, index3})
+    t.assert_equals(new_stmt_count - old_stmt_count, 15)
+    old_stmt_count = new_stmt_count
+    t.assert_equals(
+        space:select{}, {{11, 2, 13, 24, 15, 100},
+                         {2, 3, 4, 5, 6},
+                         {3, 4, 5, 6, 7},
+                         {4, 5, 6, 7, 8}})
+    t.assert_equals(
+        index2:select{}, {{2, 3, 4, 5, 6},
+                         {3, 4, 5, 6, 7},
+                         {4, 5, 6, 7, 8},
+                         {11, 2, 13, 24, 15, 100}})
+    t.assert_equals(
+        index3:select{}, {{2, 3, 4, 5, 6},
+                         {3, 4, 5, 6, 7},
+                         {4, 5, 6, 7, 8},
+                         {11, 2, 13, 24, 15, 100}})
+    -- optimize one 'secondary' index update
+    t.assert_equals(
+        index:update({3}, {{'+', 1, 10}, {'-', 5, 2}, {'!', 6, 100}}),
+        {12, 3, 4, 5, 4, 100}
+    ) -- change only index 'third'
+    t.assert_equals(box.snapshot(), 'ok')
+    dump_count = wait_for_dump(index, dump_count)
+    new_stmt_count = dumped_stmt_count({index, index2, index3})
+    t.assert_equals(new_stmt_count - old_stmt_count, 3)
+    old_stmt_count = new_stmt_count
+    -- optimize one 'third' index update
+    t.assert_equals(
+    index:update(
+        {3}, {{'=', 1, 20}, {'+', 3, 5}, {'=', 4, 30}, {'!', 6, 110}}),
+        {20, 3, 9, 30, 4, 110, 100}
+    ) -- change only index 'secondary'
+    t.assert_equals(box.snapshot(), 'ok')
+    dump_count = wait_for_dump(index, dump_count)
+    new_stmt_count = dumped_stmt_count({index, index2, index3})
+    t.assert_equals(new_stmt_count - old_stmt_count, 3)
+    old_stmt_count = new_stmt_count
+    -- optimize both indexes
+    t.assert_equals(
+        index:update({3}, {{'+', 1, 10}, {'#', 6, 1}}), {30, 3, 9, 30, 4, 100}
+    ) -- don't change any indexed fields
+    t.assert_equals(box.snapshot(), 'ok')
+    dump_count = wait_for_dump(index, dump_count)
+    new_stmt_count = dumped_stmt_count({index, index2, index3})
+    t.assert_equals(new_stmt_count - old_stmt_count, 1)
+    old_stmt_count = new_stmt_count
+    t.assert_equals(
+        space:select{}, {{11, 2, 13, 24, 15, 100},
+                         {30, 3, 9, 30, 4, 100},
+                         {3, 4, 5, 6, 7},
+                         {4, 5, 6, 7, 8}})
+    t.assert_equals(
+        index2:select{}, {{3, 4, 5, 6, 7},
+                          {4, 5, 6, 7, 8},
+                          {11, 2, 13, 24, 15, 100},
+                          {30, 3, 9, 30, 4, 100}})
+    t.assert_equals(
+        index3:select{}, {{30, 3, 9, 30, 4, 100},
+                          {3, 4, 5, 6, 7},
+                          {4, 5, 6, 7, 8},
+                          {11, 2, 13, 24, 15, 100}})
+end
+g.before_test('test_field_greater_64', function()
```

Comment:
I don't understand why some tests have the before_test action that creates indexes and populates the test space while others do it directly in the test body. For me it's okay either way though so up to you.

Reply:
I vote up the idea of consistency.

I don't like that we need to set some values into `_G` (globals) or `g` (test group object) in order to pass some data from `before_test()` to the test itself. It would be good to return values from the before_test hook and accept them as arguments in the test function.

The alternative: include the preparation code into the test itself. It looks okay too.